### ISSUE TITLE
Mistaken warning when working around `E0386` restrictions regarding mutable references in immutable containers.

### ISSUE THREAD

**Issue by nathankleyn** [rust-lang/rust#27756]
While writing the detailed diagnostics for `E0386` (cannot assign to data in an immutable container, #24407), I came across a possible mistaken warning.

`E0386` is about forbidding mutation of a mutable reference stored in an immutable container:

```rust
let mut x: i64 = 1;
let y: Box<_> = Box::new(&mut x);
**y = 2; // error, cannot assign to data in an immutable container
```

But if you make `y` mutable to get around this issue, you get a warning:

```rust
let mut x: i64 = 1;
let mut y: Box<_> = Box::new(&mut x); // warning, variable does not need to be mutable
**y = 2;
```

Although it _is_ strictly true that I don't mutate `y` itself, that I need `y` to be `mut` to be able to mutate the members makes this warning a bit dubious.

**Comment by eefriedman** [comment]
This is fixed in beta.

**Comment by nathankleyn** [comment]
@eefriedman Nice! I'll close this as I confirmed it's fine in beta.
